scene/main/node.cpp

```diff
@@ -1363,6 +1368,12 @@ void Node::_add_child_nocheck(Node *p_child, const StringName &p_name, InternalM
 
 	p_child->data.parent = this;
 
+	if (p_child->data.owner && !p_child->data.owner->is_ancestor_of(p_child)) {
+		// Owner of p_child must be ancestor of p_child.
+		WARN_PRINT(vformat("Adding '%s' as child to '%s' would make owner '%s' inconsistent. Erasing owner. Consider removing the owner beforehand.", p_child->get_name(), get_name(), p_child->data.owner->get_name()));
+		p_child->_clean_up_owner();
+	}
+
```

Comment:
I don't think this scenario should ever be possible. If it's possible, guards are missing elsewhere. But if you want to be extra sure, maybe this should only happen in editor builds?

Reply:
I'd move this check to `add_child()` instead, which is anyway the only caller of `_add_child_nocheck()` (possibly, to a counterpart of `_validate_child_name()` named `_validate_child_future_ownership(node, intended_new_parent)`). That way, `_add_child_nocheck()` still honors its name (althought the _no check_ part was about names, but now we're validating owners, it's good to extend that notion).

Aside, I think that with this we would be catching every possibility of a wrong owner. That said, given that this _may_ have a performance impact, I see value in restricting the check to debug builds, and making it just a check. I mean, keep the owner (even if invalid), not to change behavior across debug and non-debug. After all, I don't think owner chain integrity is an internal requisite of the engine (engine won't crash or misbehave due to an inconsitency there; only the user may get unexpected results).

Finally, I'd suggest changing the wording so _erasing/removing owner_ is expressed as _unsetting the owner_ instead, for enhanced clarity.

Reply:
I have moved the validation of the owner to `add_child()` within a `DEBUG_ENABLED` section.

Please note that `_add_child_nocheck()` is additionally called from `SceneState::instantiate`. But since this function deals with instantiating nodes, my guess would be, that it's unnecessary to add checks there.

I would prefer to completely prohibit inconsistent states (so that the user no longer gets unexpected results), but based on the performance-argument, I assume that a warning can be considered sufficient.

For good measure, I have also checked, that the MRP in the linked issue still crashes in v4.2.rc.custom_build [80de898d7] and gets fixed by this PR.